src/src/com/microsoft/aad/adal/AuthenticationContext.java

```diff
@@ -1430,23 +1431,31 @@ private AuthenticationResult localFlow(CallbackHandler callbackHandle,
 
         // Trying to find refresh token first, if existed, will try to use the refresh token. 
         Logger.v(TAG, "Checking refresh tokens");
-        final boolean useMultiResourceRefreshToken = cachedItem == null || cachedItem.getIsMultiResourceRefreshToken();
-        final RefreshItem refreshItem = getRefreshToken(request, useMultiResourceRefreshToken);
+        final RefreshItem refreshItem = getRefreshToken(request);
         AuthenticationResult authResult = null;
         if (!promptUser(request.getPrompt()) && refreshItem != null
                 && !StringExtensions.IsNullOrBlank(refreshItem.mRefreshToken)) {
-            Logger.v(TAG, "Refresh token is available and it will attempt to refresh token");
+            Logger.v(TAG, "Refresh token is available and will attempt to used it for a new access token.");
             try {
                 authResult = getTokenWithRefreshTokenAndUpdateCache(request, refreshItem);
             } catch (AuthenticationException authenticationException) {
-                ClientAnalytics.logEvent(
-                        InstrumentationIDs.REFRESH_TOKEN_REQUEST_FAILED,
-                        new InstrumentationPropertiesBuilder(request, authenticationException).build());
+                addLogEventForRequestFailureWithCause(InstrumentationIDs.REFRESH_TOKEN_REQUEST_FAILED, request, authenticationException);
                 callbackHandle.onError(authenticationException);
                 return null;
             }
             
-            if (authResult != null && !StringExtensions.IsNullOrBlank(authResult.getAccessToken())) {
+            // retry with MRRT if the refresh token request was done with FRT
+            if (!isAccessTokenReturned(authResult) && refreshItem != null) {
+                try {
+                    authResult = fallbackToMRRT(request, refreshItem);
```

Comment:
I would still suggest to consider the way that I mentioned to you:
refreshtoken = getFRT()
if(refreshToken != null) authresult = networkCall()
if (authresult == null or no AT) refreshToken = getRefreshToken()
if(refreshToken != null) authresult = networkCall() 
where getRefreshToken  is method that currently exist already (gets RT or MRRT and doesn't know about FRT)